tests/utils.py

```diff
@@ -0,0 +1,250 @@
+    log_dir = os.path.join(SCRIPT_DIR, "logs")
+    if not os.path.exists(log_dir):
+        os.makedirs(log_dir)
+
+    # Write stdout log for test
+    log_path = os.path.join(log_dir, "ansible_" + test_name + ".log")
+    with open(log_path, "w") as log_file:
+        log_file.write(result.stdout.decode("utf-8"))
+
+    # Write stderr log for test
+    error_log_path = os.path.join(log_dir, test_name + "-error.log")
+    with open(error_log_path, "w") as log_file:
+        log_file.write(result.stderr.decode("utf-8"))
+
+
+def _run_playbook(playbook):
+    """
+    Create a inventory using a temporary file and run ansible using it.
+
+    The logs of the run will be placed in `tests/logs/`.
+    """
+    with tempfile.NamedTemporaryFile() as inventory_file:
+        inventory_file.write(get_inventory_content())
+        inventory_file.flush()
+        cmd = [
+            "ansible-playbook",
+            "-i",
+            inventory_file.name,
+            playbook,
+        ]
+        process = subprocess.run(
+            cmd, cwd=SCRIPT_DIR, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+    test_name = get_test_name_from_playbook_path(playbook)
+    write_logs(process, test_name)
+
+    return process
+
+
+def run_playbook(playbook, allow_failures=False):
+    """
+    Run an Ansible playbook and assert the return code.
+
+    Call ansible (using _run_playbook function) and assert the result of
+    the execution.
+
+    In case of failure the tail of the error message will be displayed
+    as an assertion message.
+
+    The full log of the execution will be available in the directory
+    `tests/logs/`.
+    """
+    result = _run_playbook(playbook)
+
+    if allow_failures:
+        return result
+
+    status_code_msg = "ansible-playbook return code: {}".format(
+        result.returncode
+    )
+    assert_msg = "\n".join(
+        [
+            "",
+            "-" * 30 + " Captured stdout " + "-" * 30,
+            result.stdout.decode("utf8"),
+            "-" * 30 + " Captured stderr " + "-" * 30,
+            result.stderr.decode("utf8"),
+            "-" * 30 + " Playbook Return Code " + "-" * 30,
+            status_code_msg,
+        ]
+    )
+
+    # Need to get the last bytes of msg otherwise Azure
+    #   will cut it out.
+    assert result.returncode == 0, assert_msg[-2500:]
+
+    return result
+
+
+def list_test_yaml(dir_path):
+    """
+    List the test playbooks inside a given directory.
+
+    A test playbook is any file inside the directory which the name starts with
+    `test_` and the extension is `.yml`.
+    """
+    yamls = []
+    for yaml_name in sorted(os.listdir(dir_path)):
+        if yaml_name.startswith("test_") and yaml_name.endswith(".yml"):
+            yamls.append(
+                {
+                    "path": os.path.join(dir_path, yaml_name),
+                    "name": yaml_name.split(".")[0],
+                }
+            )
```

Comment:
Using the changes on PR #375 would also fix the execution of playbook tests in subdirectories, and would fix issue #359.

Reply:
Ok. I'll integrate your fix here.